src/common/src/hash/consistent_hash/compat.rs

```diff
@@ -15,38 +15,73 @@
+/// Directly naming it `vnode_count` will lead to the method `vnode_count()` returning
+/// `0` when the field is unset, which can be misleading sometimes.
+///
+/// Instead, we name the field as `maybe_vnode_count` and provide the method `vnode_count`
+/// through this trait, ensuring that backward compatibility is handled properly.
 pub trait VnodeCountCompat {
-    /// Returns the vnode count, or [`VirtualNode::COUNT_FOR_COMPAT`] if the vnode count is not set,
-    /// typically for backward compatibility.
+    /// Returns the vnode count if it's set. Otherwise, returns [`VirtualNode::COUNT_FOR_COMPAT`]
+    /// for distributed tables/fragments, and `1` for singleton tables/fragments, for backward
+    /// compatibility.
     ///
     /// See the documentation on the field of the implementing type for more details.
     fn vnode_count(&self) -> usize;
 }
 
-/// Implement the trait for given types by delegating to the `maybe_vnode_count` field.
-///
-/// The reason why there's a `maybe_` prefix is that, a getter method with the same name
-/// as the field will be generated for `prost` structs. Directly naming it `vnode_count`
-/// will lead to the method `vnode_count()` returning `0` when the field is unset, which
-/// can be misleading sometimes.
-///
-/// Instead, we name the field as `maybe_vnode_count` and provide the method `vnode_count`
-/// through this trait, ensuring that backward compatibility is handled properly.
-macro_rules! impl_maybe_vnode_count_compat {
-    ($($ty:ty),* $(,)?) => {
-        $(
-            impl VnodeCountCompat for $ty {
-                fn vnode_count(&self) -> usize {
-                    self.maybe_vnode_count
-                        .map_or(VirtualNode::COUNT_FOR_COMPAT, |v| v as _)
-                }
-            }
-        )*
-    };
+impl VnodeCountCompat for risingwave_pb::meta::table_fragments::Fragment {
+    fn vnode_count(&self) -> usize {
+        use risingwave_pb::meta::table_fragments::fragment::FragmentDistributionType;
+
+        if let Some(vnode_count) = self.maybe_vnode_count {
+            return vnode_count as _;
+        }
+
+        // Compatibility: derive vnode count from distribution.
+        match self.distribution_type() {
+            FragmentDistributionType::Unspecified => unreachable!(),
+            FragmentDistributionType::Single => 1,
+            FragmentDistributionType::Hash => VirtualNode::COUNT_FOR_COMPAT,
+        }
+    }
+}
+
+impl VnodeCountCompat for risingwave_pb::catalog::Table {
+    fn vnode_count(&self) -> usize {
+        if let Some(vnode_count) = self.maybe_vnode_count {
+            return vnode_count as _;
+        }
+
+        // Compatibility: derive vnode count from distribution.
+        if self.distribution_key.is_empty()
+            && self.dist_key_in_pk.is_empty()
+            && self.vnode_col_index.is_none()
+        {
+            // Singleton table.
+            1
```

Comment:
Not sure if this is strictly correct. 🤔 cc @st1page @chenzl25

Reply:
Found a counterexample: the state table for `Source` or `SourceBackfill`. Distribution there is somewhat chaotic. 🥵

Reply:
As far as I know:

1. File source is a bit special. The FS source consists of 2 executors: `FsList` ->  `FsFetch`, where `FsList` is a singleton but `FsFetch` is distributed according to vnodes.

https://github.com/risingwavelabs/risingwave/blob/911578c0a8700ee7a284480dacb8f8e07aa510d0/src/stream/src/executor/source/state_table_handler.rs#L54-L72


2. The rest sources use vnode 0 to store offsets of all splits. Meanwhile, the split manager ensures the ownership of actors-splits don't overlap with each other.

https://github.com/risingwavelabs/risingwave/blob/911578c0a8700ee7a284480dacb8f8e07aa510d0/src/stream/src/executor/source/state_table_handler.rs#L99-L103

Any other cases?

Reply:
I believe Distribution Singleton can derive both distribution_key and dist_key_in_pk, but the reverse might not be consistently maintained, as we've observed. It's likely that these inconsistencies arise from our custom-built special state tables.

Reply:
> but the reverse might not be consistently maintained, as we've observed. It's likely that these inconsistencies arise from our custom-built special state tables.

If so, then it could be hard to get this PR merged. 🥵 Or we may try assigning a vnode count of 1 _only_ for future singletons, leading to increased chaos though.

Reply:
> but the reverse might not be consistently maintained, as we've observed. It's likely that these inconsistencies arise from our custom-built special state tables.

After further investigations, I believe it's still held, as we have exactly the same code logic in the read/write interfaces when computing vnodes of a given row.

https://github.com/risingwavelabs/risingwave/blob/98a68d37e6b86a8f158982cb558f90deacb99b67/src/common/src/hash/table_distribution.rs#L70-L90

As a result, empty `dist_key` and unset `vnode_col_idx_in_pk` will lead to `ComputeVnode::Singleton` here, which always emits `0` (`SINGLETON_VNODE`) when computing vnode for rows later.